replicators/src/mysql_connector/snapshot.rs

```diff
@@ -665,6 +680,41 @@ impl MySqlReplicator {
 
         Ok(())
     }
+
+    /// This function drops all known tables from Noria. This is used when we require a full
+    /// re-snapshot of the database.
+    ///
+    /// # Arguments
+    /// * `noria`: The target ReadySet deployment
+    async fn drop_all_tables(
+        &mut self,
+        noria: &mut readyset_client::ReadySetHandle,
+    ) -> ReadySetResult<()> {
+        let replication_offsets = noria.replication_offsets().await?;
+        let mut tx = self.pool.start_transaction(tx_opts()).await?;
+        let all_tables = get_table_list(&mut tx, TableKind::BaseTable).await?;
+        let mut changes = Vec::new();
+        for (db, table) in all_tables.iter() {
+            let table = Relation {
+                schema: Some(db.clone().into()),
+                name: table.clone().into(),
+            };
+            if replication_offsets.has_table(&table) {
+                changes.push(Change::Drop {
+                    name: table,
+                    if_exists: true,
+                });
+            }
+        }
+        tx.rollback().await?;
+        noria
+            .extend_recipe_no_leader_ready(ChangeList::from_changes(
+                changes,
+                Dialect::DEFAULT_MYSQL,
+            ))
+            .await?;
+        Ok(())
```

Comment:
so the main point of the schema replication offset is to give us a logical timestamp for changes to the database that can't be represented in updates to replication offsets in tables - for example `DROP TABLE` events (since there's no table to put the new offset on!). Just thinking through this, my knee-jerk reaction is that we probably want to leave it in place (so that if a user dropped a table between the last replicated event and when we resnapshot, we replicate that `DROP TABLE` event)... 

but... I guess the offset has already overflowed? are we resetting it back to `0` as of the resnapshot? if so, we likely need to also drop any tables that *aren't* in the upstream database in addition to ones that *are*... @altmannmarcelo?

Reply:
>so that if a user dropped a table between the last replicated event and when we resnapshot, we replicate that DROP TABLE event

Won't we just set a new schema replication offset when we finish with the snapshot anyway?

If that `DROP TABLE` happens after we replicate the last event and before we resnapshot, I think that DDL change would be reflected in the new snapshot we take, since the table was dropped before we started any transactions/acquired any locks. If that `DROP TABLE` happened after our snapshotting process already began, I think we'd replicate that event after our resnapshot completes

Reply:
There is a possibility for a dangling table here I think. Not for the overflow issue, as a DDL [implicitly commits](https://dev.mysql.com/doc/refman/8.0/en/implicit-commit.html) the transaction, which will cause the binlog to rotate and the DDL will be in the new binlog. However, there is the following situation:
```
* RS is replication
* MySQL> CREATE TABLE tb1 (ID INT);
* STOP RS
* MYSQL> DROP TABLE tb1;
* MYSQL> FLUSH BINARY LOGS;
* MYSQL> PURGE BINARY LOGS TO 'current binlog'
* START RS
```

RS will try to request the old binlog which is gone and will trigger a full resnapshot. When doing the drop, it will query the MySQL database to discover the tables, but `tb1` is gone, so it will not attempt to drop it from RS.
Ideally, we should have a way ( I still need to check this) to grab the current list of tables from RS and issue a drop from this list instead of the upstream database.

Reply:
> Ideally, we should have a way ( I still need to check this) to grab the current list of tables from RS and issue a drop from this list instead of the upstream database.

you could enumerate the tables in the replication offsets, perhaps?

There's also [`Controller::tables`](https://github.com/readysettech/readyset/blob/main/readyset-client/src/controller.rs#L438)

Reply:
Oooo nice catch! [This comment](https://github.com/readysettech/readyset/blob/da5a8d5c0492a3b2a73ee8a04a6a876d99a0870f/replicators/src/noria_adapter.rs#L401-L404) suggests that if we request to start streaming the binlog from a position that doesn't exist, we'll currently error out anyway. I confirmed this by hardcoding a start position:
```
2023-09-06T20:04:22.002503Z  INFO replicators::mysql_connector::connector: Starting binlog replication next_position=ethan.0001:19329
2023-09-06T20:04:22.003314Z  WARN replicators::noria_adapter: Restarting adapter after error encountered error=Error during replication: Server error: `ERROR HY000 (1236): Could not find first log file name in binary log index file'
2023-09-06T20:04:22.003373Z ERROR replicators: Error in replication, will retry after timeout error=Error during replication: Server error: `ERROR HY000 (1236): Could not find first log file name in binary log index file' timeout_sec=1
2023-09-06T20:04:22.926093Z  INFO readyset_server::worker: worker informed of new controller controller_uri=http://127.0.0.1:6033/
2023-09-06T20:04:23.020282Z  INFO replicators::mysql_connector::connector: Starting binlog replication next_position=ethan.0001:19329
2023-09-06T20:04:23.021371Z  WARN replicators::noria_adapter: Restarting adapter after error encountered error=Error during replication: Server error: `ERROR HY000 (1236): Could not find first log file name in binary log index file'
2023-09-06T20:04:23.021409Z ERROR replicators: Error in replication, will retry after timeout error=Error during replication: Server error: `ERROR HY000 (1236): Could not find first log file name in binary log index file' timeout_sec=1
2023-09-06T20:04:24.062821Z  INFO replicators::mysql_connector::connector: Starting binlog replication next_position=ethan.0001:19329
2023-09-06T20:04:24.071553Z  WARN replicators::noria_adapter: Restarting adapter after error encountered error=Error during replication: Server error: `ERROR HY000 (1236): Could not find first log file name in binary log index file'
2023-09-06T20:04:24.071655Z ERROR replicators: Error in replication, will retry after timeout error=Error during replication: Server error: `ERROR HY000 (1236): Could not find first log file name in binary log index file' timeout_sec=1
```
In the example you gave, upon starting back up, all the tables would have replication offsets already (presumably), so we wouldn't resnapshot and would just use the max offset across all the tables in ReadySet as our binlog streaming start position. Since all the logs were purged, we'd end up requesting a position that no longer exists, so I think we'd error out.

Unfortunately, we just retry infinitely here -- the correct thing to do (as stated in the comment) would be to perform a full resnapshot if MySQL doesn't have the binlog files required to replicate events starting at our current position

Reply:
> Unfortunately, we just retry infinitely here 

Correct, We still have the offsets of the dangling table, and we repeatedly report that to upstream.

> The correct thing to do (as stated in the comment) would be to perform a full resnapshot if MySQL doesn't have the binlog files required to replicate events starting at our current position

This has been fixed on this PR [here](https://github.com/altmannmarcelo/readyset/commit/96731a547193fc1f1f219b0d001695c19d21a9b8#diff-9c7a0a9b1550ac620a8070d79c001fc8a3898db3fbc7314c299002d10f46cb5bR140). The problem was indeed the dangling table and its position not being removed. 

Removing the tables based on @glittershark 's suggestion fixes the problem.

Reply:
```
$ ./target/debug/readyset --standalone --database-type=mysql --upstream-db-url=mysql://root:@127.0.0.1:3310 --username=root --address=0.0.0.0:3307 --deployment=MySQL8.0.1 --allow-unauthenticated-connections &
[1] 1101818

$ 2023-09-07T10:02:43.439181Z  INFO readyset: Starting ReadySet adapter options=Options { address: Some(0.0.0.0:3307), deployment: "MySQL8.0.1", database_type: Some(MySQL), standalone: true, authority: Standalone, authority_address: ".", log_slow: false, allow_unauthenticated_connections: true, query_caching: Explicit, max_processing_minutes: 15, migration_task_interval: 20000, metrics_address: 0.0.0.0:6034, username: Some("root"), password: None, prometheus_metrics: false, noria_metrics: false, query_log: false, query_log_ad_hoc: false, external_address: None, tracing: Options { log_format: Full, no_color: false, log_level: "info", tracing_host: None, tracing_sample_percent: Percent(0.01), statement_logging: false, statement_log_path: None }, psql_options: Options { readyset_identity_file: None, readyset_identity_file_password: None, postgres_authentication_method: ScramSha256 }, allow_unsupported_set: false, unsupported_set_mode: Error, views_polling_interval: 5, migration_request_timeout_ms: 1800000, controller_request_timeout_ms: 5000, query_max_failure_seconds: 9223372036854775, fallback_recovery_seconds: 0, non_blocking_reads: false, embedded_readers: false, server_worker_options: WorkerOptions { durability: Permanent, persistence_threads: 6, memory_limit: 0, memory_check_freq: 1, eviction_kind: LRU, no_partial: false, allow_full_materialization: false, enable_packet_filters: false, min_workers: 1, shards: 0, volume_id: None, enable_experimental_topk_support: false, enable_experimental_paginate_support: false, enable_experimental_mixed_comparisons: false, db_dir: None, domain_replication_options: ReplicationOptions { reader_replicas: None, non_base_replicas: None }, replicator_config: UpstreamConfig { upstream_db_url: Some("mysql://root:@127.0.0.1:3310"), disable_upstream_ssl_verification: false, ssl_root_cert: None, disable_setup_ddl_replication: false, replication_server_id: None, replicator_restart_timeout: 1s, replication_tables: None, replication_tables_ignore: None, snapshot_report_interval_secs: 30, replication_pool_size: 50 }, worker_request_timeout_seconds: 1800 }, disable_telemetry: false, wait_for_failpoint: false, experimental_placeholder_inlining: false, no_upstream_connections: false, cleanup: false, controller_address: None }
 . . .
2023-09-07T10:02:44.577299Z  INFO replicators::mysql_connector::connector: Starting binlog replication next_position=binlog.000032:556
2023-09-07T10:02:44.577558Z  INFO replicators::noria_adapter: MySQL connected
2023-09-07T10:02:44.577573Z  INFO replicators::noria_adapter: binlog_position=binlog.000032:556

$ mysql -e "CREATE TABLE test.tb1 (ID INT)"

$ mysql -e "SHOW READYSET TABLES" -P 3307
+----------------+-------------+
| table          | status      |
+----------------+-------------+
| `test`.`cars`  | Snapshotted |
| `test`.`tb1`   | Snapshotted |
| `test`.`users` | Snapshotted |
+----------------+-------------+

$ kill 1101818
2023-09-07T10:03:09.766692Z  INFO RS server Shutting down: readyset: Waiting up to 20 seconds for all background tasks to shut down
. . .
2023-09-07T10:03:09.770770Z  INFO RS server Shutting down: readyset: Shutdown completed successfully
[1]+  Done                    ./target/debug/readyset --standalone --database-type=mysql --upstream-db-url=mysql://root:@127.0.0.1:3310 --username=root --address=0.0.0.0:3307 --deployment=MySQL8.0.1 --allow-unauthenticated-connections

$ mysql -e "DROP TABLE test.tb1; FLUSH BINARY LOGS; SELECT SLEEP(1); PURGE BINARY LOGS BEFORE NOW();"
+----------+
| SLEEP(1) |
+----------+
|        0 |
+----------+

$ ./target/debug/readyset --standalone --database-type=mysql --upstream-db-url=mysql://root:@127.0.0.1:3310 --username=root --address=0.0.0.0:3307 --deployment=MySQL8.0.1 --allow-unauthenticated-connections &
[1] 1102338

$ 2023-09-07T10:04:18.489596Z  INFO readyset: Starting ReadySet adapter options=Options { address: Some(0.0.0.0:3307), deployment: "MySQL8.0.1", database_type: Some(MySQL), standalone: true, authority: Standalone, authority_address: ".", log_slow: false, allow_unauthenticated_connections: true, query_caching: Explicit, max_processing_minutes: 15, migration_task_interval: 20000, metrics_address: 0.0.0.0:6034, username: Some("root"), password: None, prometheus_metrics: false, noria_metrics: false, query_log: false, query_log_ad_hoc: false, external_address: None, tracing: Options { log_format: Full, no_color: false, log_level: "info", tracing_host: None, tracing_sample_percent: Percent(0.01), statement_logging: false, statement_log_path: None }, psql_options: Options { readyset_identity_file: None, readyset_identity_file_password: None, postgres_authentication_method: ScramSha256 }, allow_unsupported_set: false, unsupported_set_mode: Error, views_polling_interval: 5, migration_request_timeout_ms: 1800000, controller_request_timeout_ms: 5000, query_max_failure_seconds: 9223372036854775, fallback_recovery_seconds: 0, non_blocking_reads: false, embedded_readers: false, server_worker_options: WorkerOptions { durability: Permanent, persistence_threads: 6, memory_limit: 0, memory_check_freq: 1, eviction_kind: LRU, no_partial: false, allow_full_materialization: false, enable_packet_filters: false, min_workers: 1, shards: 0, volume_id: None, enable_experimental_topk_support: false, enable_experimental_paginate_support: false, enable_experimental_mixed_comparisons: false, db_dir: None, domain_replication_options: ReplicationOptions { reader_replicas: None, non_base_replicas: None }, replicator_config: UpstreamConfig { upstream_db_url: Some("mysql://root:@127.0.0.1:3310"), disable_upstream_ssl_verification: false, ssl_root_cert: None, disable_setup_ddl_replication: false, replication_server_id: None, replicator_restart_timeout: 1s, replication_tables: None, replication_tables_ignore: None, snapshot_report_interval_secs: 30, replication_pool_size: 50 }, worker_request_timeout_seconds: 1800 }, disable_telemetry: false, wait_for_failpoint: false, experimental_placeholder_inlining: false, no_upstream_connections: false, cleanup: false, controller_address: None }
. . .
2023-09-07T10:04:19.584299Z  INFO replicators::mysql_connector::connector: Starting binlog replication next_position=binlog.000032:747
2023-09-07T10:04:19.584462Z ERROR replicators::mysql_connector::connector: Failed to request binlog error=ERROR HY000 (1236): Could not find first log file name in binary log index file
2023-09-07T10:04:22.585867Z  WARN replicators::noria_adapter: Restarting adapter after error encountered. Full resnapshot will be performed error=Fatal error requires full resnapshot
2023-09-07T10:04:22.591493Z  INFO taking database snapshot: replicators::noria_adapter: Starting snapshot
. . .
2023-09-07T10:04:23.240306Z  INFO taking database snapshot: replicators::noria_adapter: Snapshot finished
2023-09-07T10:04:23.244975Z  INFO replicators::mysql_connector::connector: Starting binlog replication next_position=binlog.000033:157
2023-09-07T10:04:23.245168Z  INFO replicators::noria_adapter: MySQL connected
2023-09-07T10:04:23.245184Z  INFO replicators::noria_adapter: binlog_position=binlog.000033:157

$ mysql -e "SHOW READYSET TABLES" -P 3307
+----------------+-------------+
| table          | status      |
+----------------+-------------+
| `test`.`cars`  | Snapshotted |
| `test`.`users` | Snapshotted |
+----------------+-------------+

```